src/credentials/loginManager.ts

```diff
@@ -227,3 +238,125 @@ function tryMakeCredentialsProviderId(credentials: string): CredentialsId | unde
         return undefined
     }
 }
+
+async function getProvider(id: CredentialsId): Promise<CredentialsProvider> {
+    const provider = await CredentialsProviderManager.getInstance().getCredentialsProvider(id)
+    if (!provider) {
+        throw new Error(`Could not find Credentials Provider for ${asString(id)}`)
+    }
+    return provider
+}
+
+/**
+ * The Toolkit implementation has a good amount of custom logic (SSO, source profiles, etc.)
+ * that was written to fill feature gaps in both AWS SDK V2 and V3. This code works imperfectly with
+ * pre-existing SDK refresh logic, leading to users experiencing issues with credentials expiring and
+ * forcing them to re-select a profile to refresh.
+ *
+ * So, this interface sits between everything else to mediate the refreshes. Adding a thin interface
+ * is preferred over bulking up existing ones since it allows for clean(-ish) refactors against the
+ * credentials subsystem. The pure data structure shape that existed previously was better, but it
+ * just wouldn't be able to support refreshes on its own.
+ */
+export interface CredentialsShim {
+    /**
+     * Fetches credentials, attempting a passive refresh if needed.
+     *
+     * Passive refreshes will automatically fail when user input is required.
+     */
+    get: () => Promise<Credentials>
+
+    /**
+     * Removes the stored credentials and performs an active refresh, allowing for prompts.
+     *
+     * Calling this function while a refresh is still pending returns the already pending promise.
+     */
+    refresh: () => Promise<Credentials>
```

Comment:
May drop this method. Only have it because SDK v2 differentiates between getting and refreshing while v3 does not.

Reply:
Still needed because the SDK hides refresh logic internally.